### PR TITLE
Added .copy() to Layer and Stack

### PR DESCRIPTION
**Pull request by TenebraeX8**
Now you can make copies of Stacks and layers. This is the way you can do it without including sklearn or any other dependencies

### CODE REVIEW

**Comment by Cheukting** [comment]
Hi @TenebraeX8 Thanks for your PR. 2 points to notice here:

1. Since you have implemented a new method. We need test cases to test them. Could you please write the test cases as well?

2. I am sure your implementation will work. But for performance consideration, your implementation will make each layer store 2 sets of models and preprocessors and will double the memory used. I think it will be better if the architecture of the layers are saved as a list of string (name of the model) as an attribute and build them on demand (when `.copy()`) is called. It is just my point of view, let me know what you think about this.

Anyway, thank you for your work. Let me know if you want to continue working on this PR and adding the test cases.

**Comment by TenebraeX8** [comment]
Hi @Cheukting. Thanks for reviewing my code.

1. Of course I can do that. I was just waiting because I wasnt sure if you integrate it into the project and I saw you are writing pretty extensive tests. Gonna have a look at it most probably within the week

2. Yeah, this implementation is not performant at the first look. Still, I choose it because there is no further dependency needed and (what was the most significant point) it will work with every model. I surely could save the single types and re-construct it. BUT: Every parameter passed by the construction of the actual model (e.g. LinearRegression(normalize=True)) will be gone. So if I do that, it will not be the same model as it was before.
There is another option. The method sklearn.base.clone is shallow-copying estimators without any trainings-dependencies. I tried implementing it at first, but it would have cost the dependency to sklearn.base, what would have had an enormous impact on the project as you tried your best writing it without any dependencies to this library.
To get back to performance (the time aspect): As you already make a deep-copy of the model when constructing the layer, the performance with doing it a second time rises from O(n) to O(2n), not a huge impact. Also, when working with training networks building a structure is kind of the thing which consumes the least amount of time.
In order to resolve the memory issue, I thought of an option to let the user decide:
when constructing the layer, there will be another parameter: makeCopiable. Default value is false. If the user sets it to true, the variables are set. Otherwise not and the structure will not be able to be copied and produces a runtime-error or returns None, whatever you like best.
This way, you still have the performance of the old structure if you dont want to copy and the possibility to copy with just a bit more memory loss if you want.

What do you think?

**Comment by Cheukting** [comment]
Thank you for the explanation @TenebraeX8 . What you said is totally reasonable, in my point of view, I like the idea of using the `sklearn.base.clone` method. This tool now is mostly useful with Scikit-learn at the moment (and thus assume having Scikit-learn as a dependency) and I think for input with a list of Scikit-learn models, using the method provided by Scikit-learn will be the best way. 

When we support more models (e.g. Keras) other than Scikit-learn, we can then think about having different implementation for different inputs. Side note, Keras also have a similar function `keras.models.clone_model` 

Having the user to choose keeping a 'vanilla copy' or not would also a possible solution as only the user know if they want to 'spend' double on the memory to store a 'vanilla copy'. However, in a design point of view, I would rather this tool handle everything automatically.

It would be great if you can use the `sklearn.base.clone` like you suggest at the beginning.

**Comment by TenebraeX8** [comment]
@Cheukting Ok. I will have a look at it and use sklearn.base. as a further dependency. You will hear from me within the week.

**Comment by neomatrix369** [comment]
I think it's a good idea to also build some sort of performance measurement metrics around such functionality, either in the form of tests or a performance metric(s) as part of the functionality.

I'll give an example:


In the image above one of the meta-data is the time it took to generate that summarisation, I did this for my library so that I know which methods take up how much time - good to know from an engineering/architecture point of view. In case introduce regression or something went wrong in that particular functionality if the duration varies between different runs.

-- Just my 2p, at the end of the day please use your own discretion.

**Comment by TenebraeX8** [comment]
@Cheukting as agreed I wrote the new function. Please have a look at it and tell me, if you want any further changes

File: picknmix/picknmix.py [hunk 1]
```diff
@@ -9,6 +9,7 @@
 from copy import deepcopy
 import numpy as np
 import warnings
+import sklearn
```
Comment (Cheukting) [inline picknmix/picknmix.py]: All look great, however, to make it better, we can first check if the layer is composed of any Scikit-learn model first. You can refer to how `sklearn.base.clone` do it in their [source code](https://github.com/scikit-learn/scikit-learn/blob/1495f6924/sklearn/base.py#L32) basically they are checking if there is a `get_params` method. After that, we can import `sklearn.base.clone` by `__import__` method. I think this will also eliminate some dependency issue.

Of cause, in this case, for a non-scikit-learn object we have to make a deepcopy instead.

Comment (TenebraeX8) [reply picknmix/picknmix.py]: I will do it. using importlib (equal to __import__) sounds like a very good idea. Gimme some time and I will implement it too. Another ticket wont be neccessary

Comment (TenebraeX8) [reply picknmix/picknmix.py]: @Cheukting Done and done. Dependency is gone! :)
